### PR TITLE
Object Detection , Segmentation and Pose Detection Datasets

### PR DESCRIPTION
**Pull request by Akash-guna**
Adds Object Detection , Segmentation and Pose Detection Datasets

Previous PR #5 (#4 was closed)

### CODE REVIEW

File: core/datasets/__init__.py [hunk 1]
```diff
@@ -1,35 +1,35 @@
...
+    elif name == "SegmentationDatasetFactory":
+        from .segmentation import DatasetFactory as SegmentationDatasetFactory
+
+        return SegmentationDatasetFactory
+
+    # tracking (commented out in original)
+    # elif name == "TrackingDatasetFactory":
+    #     from trailmet.datasets.tracking import DatasetFactory as TrackingDatasetFactory
+    #     return TrackingDatasetFactory
+
```
Comment (shwu-nyunai) [inline core/datasets/__init__.py:27]: is this coded needed?

File: core/datasets/classification/cifar.py
Comment (shwu-nyunai) [inline core/datasets/classification/cifar.py:1]: the only change in this file is adding the documentation to the class. I assume this is unintentional as a previous commit removed a similar document. Ignore if intentional.

File: core/datasets/object_detection/__init__.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        if "COCODETECTION" == name:
+            obj_dfactory = CocoDetectionDataset(**kwargs)
+        elif "VOCDETECTION" == name:
+            obj_dfactory = VOCDetectionDataset(**kwargs)
+        elif "CUSTOMCOCODETECTION" == name:
+            obj_dfactory = CocoDetectionDataset(download=False, custom=True**kwargs)
+        elif "CUSTOMVOCDETECTION" == name:
+            obj_dfactory = VOCDetectionDataset(download=False, **kwargs)
+        else:
+            raise Exception(f"unknown dataset{kwargs['name']}")
```
Comment (shwu-nyunai) [inline core/datasets/object_detection/__init__.py:35]: nitpick -

missing a space in error message. u in unknown could be capital.

File: core/datasets/object_detection/cocodetection.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+            target_transform=target_transform,
+            download=download,
+            split_types=split_types,
+            val_fraction=val_fraction,
+            shuffle=shuffle,
+            random_seed=random_seed,
+        )
+
+        ## To do: check val_frac is float, else raise error
+        ## To do: if shuffle is true, there should be 'val' in train test split
```
Comment (shwu-nyunai) [inline core/datasets/object_detection/cocodetection.py:72]: how important are these todo's?

File: core/datasets/object_detection/cocodetection.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                    os.path.exists(
+                        os.path.join(
+                            self.custom_root, "annotations", "instances_train2017.json"
+                        )
+                    ),
+                ]
+            ):
+                self.download_mscoco(self.custom_root)
+            else:
+                pass
```
Comment (shwu-nyunai) [inline core/datasets/object_detection/cocodetection.py:121]: else block is not needed

File: core/datasets/object_detection/cocodetection.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+            download_and_extract_archive(
+                "http://images.cocodataset.org/zips/train2017.zip", to_path
+            )
+            download_and_extract_archive(
+                "http://images.cocodataset.org/zips/val2017.zip", to_path
+            )
+            download_and_extract_archive(
+                "http://images.cocodataset.org/annotations/annotations_trainval2017.zip",
+                to_path,
+            )
```
Comment (shwu-nyunai) [inline core/datasets/object_detection/cocodetection.py:188]: if these files are large, we can make the downloads parallel?

this change can be moved to subsequent dev cycles or ft. issues. @Arnav0400  what's your opinion?

File: core/datasets/object_detection/dataset.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        # check if split_types contain nothing beyond train, val and test
+        assert (
+            len(set(self.split_types) - set(["train", "val", "test"])) == 0
+        ), "Only train, val and test permitted as inputs in split_types"
+
+        if self.shuffle:
+            # ensure that val exists in split_types
+            assert (
+                "val" in self.split_types
+            ), "Required param val missing in the defined split types (split_type)"
```
Comment (shwu-nyunai) [inline core/datasets/object_detection/dataset.py:75]: nitpick- 
I believe the assertion error message should be an f-string and the last word would be enclosed in '{}'.
ignore if intentional.

File: core/datasets/pose_estimation/cocopose.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+            target_transform=target_transform,
+            download=download,
+            split_types=split_types,
+            val_fraction=val_fraction,
+            shuffle=shuffle,
+            random_seed=random_seed,
+        )
+
+        ## To do: check val_frac is float, else raise error
+        ## To do: if shuffle is true, there should be 'val' in train test split
```
Comment (shwu-nyunai) [inline core/datasets/pose_estimation/cocopose.py:69]: similar as above, ignore if not a problem.

File: core/datasets/pose_estimation/cocopose.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+            download_and_extract_archive(
+                "http://images.cocodataset.org/zips/train2017.zip", to_path
+            )
+            download_and_extract_archive(
+                "http://images.cocodataset.org/zips/val2017.zip", to_path
+            )
+            download_and_extract_archive(
+                "http://images.cocodataset.org/annotations/annotations_trainval2017.zip",
+                to_path,
+            )
```
Comment (shwu-nyunai) [inline core/datasets/pose_estimation/cocopose.py:177]: similar concerns as objectdet

**Review by shwu-nyunai — CHANGES_REQUESTED** [review]
I see redundancies in BaseDataset classes; if possible, can we abstract out the common methods and extend/sub-class the download and modification utilities?

File: core/datasets/object_detection/cocodetection.py (hunk 5 above)
Comment (Akash-guna) [reply core/datasets/object_detection/cocodetection.py:188]: This wont be used every time, and not every user will use this. So it is currently of a lower priority. Can be implemented in future.

File: core/datasets/__init__.py (hunk 1 above)
Comment (Akash-guna) [reply core/datasets/__init__.py]: All comments mentioned in this pr are removed

File: core/datasets/object_detection/dataset.py [hunk 9]
```diff
@@ -72,7 +72,7 @@ def __init__(
             # ensure that val exists in split_types
             assert (
                 "val" in self.split_types
-            ), "Required param val missing in the defined split types (split_type)"
+            ), f"Required param val missing in the defined split types (split_type)"
```
Comment (shwu-nyunai) [inline core/datasets/object_detection/dataset.py:75]: there wasn't a need of making this an f-string if not using self.split_types in the error message 🙂

**Review by shwu-nyunai — APPROVED** [review]

**Comment by shwu-nyunai** [comment]
> I see redundancies in BaseDataset classes; if possible, can we abstract out the common methods and extend/sub-class the download, modification and other utilities?

any comments?

**Review by shwu-nyunai — APPROVED** [review]
awsm.

**Review by Arnav0400 — APPROVED** [review]

**Pull request merged**